### ISSUE TITLE
0.8794 → 0.3243 성능 급락 리그레션 원인 규명 & 복구

### ISSUE THREAD

**Issue by iejob** [iejob/computer-vision-competition-1SEN#15]
## 📋 작업 개요

**작업 주제:** 김병현 `0.87_main.ipynb` 모듈화 후 성능 급락(0.8794→0.3243) 원인 규명 및 복구

**담당자:** @iejob

**마감일:** 9/05 24:00

## 📅 기간

- 시작일: 2025-09-04
- 종료일: 2025-09-05

---

## 📌 이슈 목적

모듈화 과정에서의 **설정/데이터/변환/모델/체크포인트/추론** 불일치를 전수 점검하여 성능을 **0.87대 재현** 또는 근접 복구한다.

---

### A. 재현 기준선 확보 (노트북 vs 모듈)

- [ ]  원본 노트북 환경에서 **동일 데이터**로 1회 학습/검증 재현(홀드아웃 또는 동일 폴드)
- [ ]  모듈 버전 동일 하이퍼로 재학습 → **metrics.jsonl**의 `macro_f1` 비교
- [ ]  두 실행의 `config.yaml`/로그 핵심 키(모델명, img_size, pooling, num_classes, aug, seed) **Diff 리포트** 생성

### B. 설정(diff) 전수 비교

- [ ]  `configs/train.yaml` ↔ 노트북 하이퍼(입력 크기, optimizer/scheduler, lr/epochs/batch, label_smoothing, amp, grad_clip) **1:1 대조표** 작성
- [ ]  `configs/infer.yaml`의 `img_size/tta/ckpt.path/pooling`이 **학습 설정과 일치** 확인
- [ ]  `num_classes`, `pooling(global_pool)`, `drop_rate/drop_path_rate` 불일치 탐지 스크립트 실행

### C. 데이터/라벨 매핑 검증

- [ ]  `meta.csv` 라벨 맵과 **모델 head 순서** 일치 검증(클래스 인덱스 전수 체크)
- [ ]  `train.csv`의 `ID` 확장자 유무 혼재 시 **이중 확장자(.jpg.jpg) 방지** 및 확장자 fallback 동작 확인
- [ ]  `BGR/RGB` 채널 순서 및 `Gray→RGB` 변환 일관성 검증(샘플 128장 로드 후 텐서 통계 비교)

### D. 변환/증강(Train/Valid/Infer) 일치성

- [ ]  **학습의 검증 변환** ↔ **추론 변환** 완전 일치 확인(Resize/Pad/Normalize 파이프라인)
- [ ]  노트북에 존재했던 증강 파라미터(rotate/hflip/clahe 등)와 모듈의 기본값 **차이표** 생성
- [ ]  `img_size`가 학습/검증/추론 모두 동일한지 확인(다르면 즉시 재실험)

### E. CV/스플릿/시드

- [ ]  `StratifiedKFold(k, seed, shuffle)` 세팅 및 폴드 분배표 **동일성** 확인
- [ ]  모듈에서 `valid_fold="all"`/단일 폴드 혼용 여부 점검 → 평가 프로토콜 통일
- [ ]  시드 고정 유틸(NumPy/torch/cudnn) 호출 여부 로그로 검증

### F. 모델/체크포인트 정합성

- [ ]  `build_model(name, num_classes, pooling)`과 ckpt의 head shape **호환성** 점검
- [ ]  **실제 로드된 ckpt 경로**를 로그에서 추출(자동 탐색 vs 명시 경로)
- [ ]  `state_dict` 키 미스매치/shape mismatch 경고 로그 유무 검사

### G. 추론 경로/후처리 검증

- [ ]  TTA 각도/집계 방식(softmax 평균 vs 로짓 평균) 확인 및 노트북 동작과 일치화
- [ ]  submit CSV `ID` 정렬/중복/NaN/형식(헤더, index=False) 검증
- [ ]  단일 최고 모델 vs 폴드 앙상블 차이 비교(가능 시 둘 다 생성)

### H. 로그 기반 원인 스코핑

- [ ]  `logs/train/*.log`에서 `[CFG]/[PATH]/[MODEL]/[EPOCH]/NEW_BEST` 핵심 라인 grep 추출
- [ ]  `experiments/.../metrics.jsonl` 집계 → **best valid macro_f1**와 에폭/체크포인트 상관 확인
- [ ]  `logs/infer/*.log`에서 `[CKPT]/[OUT]/[EXIT]` 라인 점검(ckpt/출력 경로/성공 코드)

### I. 빠른 역추적 실험(원인 후보 A/B)

- [ ]  **img_size**만 노트북 값으로 고정 후 A/B
- [ ]  **pooling(=avg) 강제**, `num_classes`/head 재생성 후 A/B
- [ ]  **검증/추론 변환 완전 동일화** 후 A/B
- [ ]  **ckpt 명시 로드**(`-ckpt`) vs 자동 탐색 A/B

### J. 결과 정리 & 복구

- [ ]  원인/증상/조치/결과 요약 표 작성(스크린샷 포함)
- [ ]  **복구된 설정 파일** 커밋(`configs/train.yaml`, `configs/infer.yaml`)
- [ ]  실행 가이드 문서(Training/Inference) 수정 및 로그 링크 첨부

---

## 📦 설치/실행 명령어 예시

```bash
# 1) 환경 확인
python -V && pip list | grep -E "torch|timm|albumentations|opencv|pandas|numpy|Pillow"

# 2) 학습 재현(모듈)
python -m src.training.train_main --config configs/train.yaml

# 3) 추론 재현(모듈)
python -m src.inference.infer_main --config configs/infer.yaml \
  --out submissions/20250904/infer.csv

# 4) 로그/메트릭 핵심 추출
grep -E "^\[CFG\]|\[PATH\]|\[MODEL\]|\[EPOCH\]|NEW_BEST|\[EXIT\]" logs/train/train_*.log | tail -n 200
grep -E "^\[CKPT\]|\[OUT\]|\[EXIT\]" logs/infer/infer_*.log

# 5) metrics.jsonl에서 best macro_f1 확인
python - <<'PY'
import json,glob
paths=glob.glob("experiments/*/*/metrics.jsonl")
for p in paths:
    rows=[json.loads(l) for l in open(p)]
    best=max((r for r in rows if "macro_f1" in r), key=lambda r:r["macro_f1"], default=None)
    if best: print(p, "BEST", best.get("fold"), best["epoch"], best["macro_f1"])
PY

# 6) ckpt-모델 호환성(헤드 차원) 점검 스니펫(필요 시)
python - <<'PY'
import torch, yaml
from src.models.build import build_model
cfg=yaml.safe_load(open("experiments/20250904/v087/config.yaml"))
m=build_model(cfg["model"]["name"], cfg["data"]["num_classes"], cfg["model"].get("pretrained",True), cfg["model"].get("drop_rate",0.0), cfg["model"].get("drop_path_rate",0.0), cfg["model"].get("pooling","avg"))
sd=torch.load("experiments/20250904/v087/ckpt/best_fold0.pth", map_location="cpu")
missing, unexpected = m.load_state_dict(sd, strict=False)
print("missing:", missing[:5], "unexpected:", unexpected[:5], "…")
PY
```

---

### ⚡️ 참고

- 평가 급락의 **가장 흔한 원인**: `img_size`/정규화/검증·추론 변환 불일치, `pooling/num_classes` 헤드 불일치, 잘못된 ckpt 자동 탐색, `StratifiedKFold`/seed 차이, `RGB↔BGR` 채널/확장자 처리 오류.
- `logs/`와 `experiments/`가 이미 잘 기록되고 있으니, **로그 3점 세트**(`[CFG]/[MODEL]`, `NEW_BEST`, `[CKPT]/[OUT]`) 우선 점검 권장.
